### PR TITLE
[oneDNN v3.1] Enabled INT8 Conv op/fusions with oneDNN v3.1

### PR DESCRIPTION
**Pull request by mahmoud-abuzaina**
- This PR adds support for oneDNN v3.1 in QuantizedConvolution op/fusions.

- Main changes in oneDNN v3.0 for quantization API are 1) Scaling API; now scale needs to be set for each tensor. 2) Bias needs to be passed as float32.

- This PR is not changing common (non-oneDNN) TF code.

### CODE REVIEW

**Comment by gbaned** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the PR and sorry for the delay! I have minor comments.

File: tensorflow/core/kernels/mkl/mkl_conv_ops.cc [hunk 1]
```diff
@@ -450,16 +520,25 @@ class MklConvFwdPrimitive : public MklPrimitive {
...
+                  {DNNL_ARG_SCRATCHPAD, *context_.sp_mem},
+                  {DNNL_ARG_DST, *context_.dst_mem}};
+#ifdef ENABLE_ONEDNN_V3
+      if (is_scale_set["src"] && is_scale_set["wei"] && is_scale_set["dst"]) {
+        net_args.insert(
+            {DNNL_ARG_ATTR_SCALES | DNNL_ARG_SRC, *context_.src_scale_mem});
+        net_args.insert(
+            {DNNL_ARG_ATTR_SCALES | DNNL_ARG_WEIGHTS, *context_.wei_scale_mem});
+        net_args.insert(
+            {DNNL_ARG_ATTR_SCALES | DNNL_ARG_DST, *context_.dst_scale_mem});
```
[the remark this answers is not in the document]
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl/mkl_conv_ops.cc]: Fixed.

File: tensorflow/core/kernels/mkl/mkl_conv_ops.cc [hunk 2]
```diff
@@ -470,26 +549,35 @@ class MklConvFwdPrimitive : public MklPrimitive {
...
```
[the remark this answers is not in the document]
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl/mkl_conv_ops.cc]: Fixed.

File: tensorflow/core/kernels/mkl/mkl_conv_ops.cc [hunk 3]
```diff
@@ -2300,6 +2502,98 @@ class MklQuantizedConvOp
...
+        (min_filter_vector.shape() != max_filter_vector.shape())) {
+      TF_CHECK_OK(absl::FailedPreconditionError(
+          "`min_filter and max_filter` must have same"
+          "shape and contain at least one element."));
+    }
+    const float* min_filter = min_filter_vector.flat<float>().data();
+    const float* max_filter = max_filter_vector.flat<float>().data();
+    const float int_const_scale_limit =
+        (std::is_same<Tinput, quint8>::value) ? 255.0 * 127.0 : 127.0 * 127.0;
+    // Re-scale bias if either of following 2 conditions are met:
```
[the remark this answers is not in the document]
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl/mkl_conv_ops.cc:2530]: Added.

File: tensorflow/core/kernels/mkl/mkl_conv_ops.cc [hunk 4]
```diff
@@ -2300,6 +2502,98 @@ class MklQuantizedConvOp
...
+    // 1. Bias is not const;
+    // 2. Bias is const, bias has not been cached (first iteration).
+
```
[the remark this answers is not in the document]
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl/mkl_conv_ops.cc]: Done.

File: tensorflow/core/kernels/mkl/mkl_conv_ops.cc [hunk 5]
```diff
@@ -2300,6 +2502,98 @@ class MklQuantizedConvOp
...
+          const_cast<Tbias*>(bias_tensor.flat<Tbias>().data()));
+      if (!input_bias_) {
+        input_bias_ = new memory(bias_md, this->cpu_engine_, bias_buf);
+      } else {
+        input_bias_->set_data_handle(bias_buf);
+      }
+
+      if (!scaled_bias_buf_)
+        AllocTmpBuffer<float>(context, &scaled_bias_tensor_,
+                              conv_fwd_pd->bias_desc(), &scaled_bias_buf_);
```
[the remark this answers is not in the document]
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl/mkl_conv_ops.cc]: Done.

**Comment by mahmoud-abuzaina** [comment]
Thank you for reviewing the PR. I have addressed your comments.

**Review by penpornk — APPROVED** [review]
Thank you!

**Pull request merged**